cupy/padding/pad.py

```diff
@@ -1,321 +1,744 @@
+            pad_area = _slice_at_axis(
+                slice(-right_pad, -right_pad + period), axis
+            )
+            new_right_pad = right_pad - period
+        else:
+            # Chunk matches pad area
+            pad_area = _slice_at_axis(slice(-right_pad, None), axis)
+        padded[pad_area] = left_chunk
+
+    return new_left_pad, new_right_pad
+
+
+def _as_pairs(x, ndim, as_index=False):
+    """Broadcast `x` to an array with the shape (`ndim`, 2).
+
+    A helper function for `pad` that prepares and validates arguments like
+    `pad_width` for iteration in pairs.
+
+    Args:
+      x({None, scalar, array-like}): The object to broadcast to the shape
+          (`ndim`, 2).
+      ndim(int): Number of pairs the broadcasted `x` will have.
+      as_index(bool, optional): If `x` is not None, try to round each
+          element of `x` to an integer (dtype `cupy.intp`) and ensure every
+          element is positive. (Default value = False)
 
     Returns:
-        cupy.ndarray:
-        Padded array of rank equal to ``array`` with shape increased according
-        to ``pad_width`` .
+      nested iterables, shape (`ndim`, 2): The broadcasted version of `x`.
+    """
+    if x is None:
+        # Pass through None as a special case, otherwise cupy.round(x) fails
+        # with an AttributeError
+        return ((None, None),) * ndim
+
+    x = numpy.array(x)
+    if as_index:
+        x = numpy.asarray(numpy.round(x), dtype=numpy.intp)
+
+    if x.ndim < 3:
+        # Optimization: Possibly use faster paths for cases where `x` has
+        # only 1 or 2 elements. `numpy.broadcast_to` could handle these as well
+        # but is currently slower
+
+        if x.size == 1:
+            # x was supplied as a single value
+            x = x.ravel()  # Ensure x[0] works for x.ndim == 0, 1, 2
+            if as_index and x < 0:
+                raise ValueError("index can't contain negative values")
+            return ((x[0], x[0]),) * ndim
+
+        if x.size == 2 and x.shape != (2, 1):
+            # x was supplied with a single value for each side
+            # but except case when each dimension has a single value
+            # which should be broadcasted to a pair,
+            # e.g. [[1], [2]] -> [[1, 1], [2, 2]] not [[1, 2], [1, 2]]
+            x = x.ravel()  # Ensure x[0], x[1] works
+            if as_index and (x[0] < 0 or x[1] < 0):
+                raise ValueError("index can't contain negative values")
+            return ((x[0], x[1]),) * ndim
+
+    if as_index and x.min() < 0:
+        raise ValueError("index can't contain negative values")
+
+    # Converting the array with `tolist` seems to improve performance
+    # when iterating and indexing the result (see usage in `pad`)
+    try:
+        return numpy.broadcast_to(x, (ndim, 2)).tolist()
+    except AttributeError:
+        # NumPy < 1.10 does not have broadcast_to
+        return cupy.broadcast_to(cupy.asarray(x), (ndim, 2)).tolist()
```

Comment:
We don't want it to fallback to cupy routines for the performance.
How about rewriting `x.shape` directly
```suggestion
        x.shape = (ndim, 2)
        return x
```
or using `numpy.repeat`?
```suggestion
        return numpy.repeat(x, ndim * 2 // x.size).reshape(ndim, 2)
```
I think the former one is more appropriate.

Reply:
Setting the shape attribute seems to work and is very fast. It takes only a few hundred nanoseconds vs. a few microseconds for `np.broadcast_to` on my system.

Reply:
I couldn't think of a case where broadcast_to should work when setting the shape shouldn't, so I will just use your suggestion all NumPy versions. thanks